DSharpPlus/Entities/AuditLogParser.cs

```diff
@@ -0,0 +1,1651 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Globalization;
+using System.Linq;
+using System.Threading.Tasks;
+using DSharpPlus.Enums;
+using DSharpPlus.Net.Abstractions;
+using DSharpPlus.Net.Serialization;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json.Linq;
+
+namespace DSharpPlus.Entities;
+
+internal static class AuditLogParser
+{
+    internal static async Task<IEnumerable<DiscordAuditLogEntry>> ParseAuditLogToEntriesAsync
+    (
+        BaseDiscordClient client,
+        DiscordGuild guild,
+        AuditLog auditLog
+    )
+    {
+        //Get all User
+        IEnumerable<DiscordUser> users = auditLog.Users;
+
+        //Update cache if user is not known
+        foreach (DiscordUser discordUser in users)
+        {
+            discordUser.Discord = client;
+            if (client.UserCache.ContainsKey(discordUser.Id))
+            {
+                continue;
+            }
+
+            client.UpdateUserCache(discordUser);
+        }
+
+        //get unique webhooks, scheduledEvents, threads
+        IEnumerable<DiscordWebhook> uniqueWebhooks = auditLog.Webhooks;
+        IEnumerable<DiscordScheduledGuildEvent> uniqueScheduledEvents = auditLog.Events;
+        IEnumerable<DiscordThreadChannel> uniqueThreads = auditLog.Threads;
+        Dictionary<ulong, DiscordWebhook> webhooks = uniqueWebhooks.ToDictionary(x => x.Id);
+
+        //update event cache and create a dictionary for it 
+        Dictionary<ulong, DiscordScheduledGuildEvent> events = new();
+        foreach (DiscordScheduledGuildEvent discordEvent in uniqueScheduledEvents)
+        {
+            guild._scheduledEvents[discordEvent.Id] = discordEvent;
+        }
+
+        events = guild._scheduledEvents.ToDictionary(x => x.Key, y => y.Value);
+
+        Dictionary<ulong, DiscordThreadChannel> threads = new();
+        if (uniqueThreads.Any())
+        {
+            threads = uniqueThreads.ToDictionary(xa => xa.Id, xa => xa);
+        }
```

Comment:
do we cache threads anywhere? if so, worth updating

Reply:
is updated